shared/src/sys/mod.rs

```diff
@@ -38,3 +99,129 @@ impl<'a> TryFrom<&'a crate::econ::TokenAmount> for TokenAmount {
         })
     }
 }
+
+impl fmt::Display for TokenAmount {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{}", crate::econ::TokenAmount::from(self))
+    }
+}
+
+impl Add<Self> for TokenAmount {
+    type Output = Self;
+
+    fn add(self, rhs: Self) -> Self::Output {
+        crate::econ::TokenAmount::from(self)
+            .add(&crate::econ::TokenAmount::from(rhs))
+            .try_into()
+            .unwrap()
+    }
+}
+
+impl Sub<Self> for TokenAmount {
+    type Output = Self;
+
+    fn sub(self, rhs: Self) -> Self::Output {
+        // panic if rhs > self?
+        crate::econ::TokenAmount::from(self)
+            .sub(&crate::econ::TokenAmount::from(rhs))
+            .try_into()
+            .unwrap()
+    }
+}
+
+impl Zero for TokenAmount {
+    #[inline]
+    fn zero() -> TokenAmount {
+        TokenAmount { hi: 0, lo: 0 }
+    }
+
+    #[inline]
+    fn set_zero(&mut self) {
+        self.hi = 0;
+        self.lo = 0;
+    }
+
+    #[inline]
+    fn is_zero(&self) -> bool {
+        self.hi == 0 && self.lo == 0
+    }
+}
+
+impl Default for TokenAmount {
+    #[inline]
+    fn default() -> TokenAmount {
+        Zero::zero()
+    }
+}
+
+impl PartialEq<Self> for TokenAmount {
+    fn eq(&self, other: &Self) -> bool {
+        self.hi == other.hi && self.lo == other.lo
+    }
+}
+
+impl Eq for TokenAmount {}
+
+impl Hash for TokenAmount {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.lo.hash(state);
+        if self.hi != 0 {
+            self.hi.hash(state);
+        }
+    }
+}
+
+impl Mul<u64> for TokenAmount {
+    type Output = TokenAmount;
+
+    fn mul(self, rhs: u64) -> Self::Output {
+        let eta: crate::econ::TokenAmount = self.into();
+        crate::econ::TokenAmount::from(eta * rhs)
+            .try_into()
+            .unwrap()
```

Comment:
We don't actually _need_ BigInt here. We can:

1. Convert to a u128 (this should compile to a cast if we do it right).
2. Perform a checked operation with u128.
3. Convert back.

Reply:
Yessir!